artichoke-backend/src/extn/core/time/subsec.rs

```diff
@@ -0,0 +1,557 @@
+use crate::convert::implicitly_convert_to_int;
+use crate::extn::core::symbol::Symbol;
+///! Parser of Ruby Time subsecond parameters to help generate `Time`.
+///!
+///! This module implements the logic to parse two optional parameters in the
+///! `Time.at` function call. These parameters (if specified) provide the number
+///! of subsecond parts to add, and a scale of those subsecond parts (millis, micros,
+///! and nanos).
+use crate::extn::prelude::*;
+
+const NANOS_IN_SECOND: i64 = 1_000_000_000;
+
+const MILLIS_IN_NANO: i64 = 1_000_000;
+const MICROS_IN_NANO: i64 = 1_000;
+const NANOS_IN_NANO: i64 = 1;
+
+#[allow(clippy::cast_precision_loss)]
```

Comment:
There is a not insignificant amount of casting going on in this class - and I'm not sure if it's avoidable. I'm throwing a comment here, because I'm sure there are better solutions than what I've done here, so I'm all :ear:s

Reply:
Unfortunately there's not a much better way to do this on stable today with "checked"-style APIs.

There are _unchecked_ APIs:

- [`f64::to_int_unchecked`](https://doc.rust-lang.org/stable/std/primitive.f64.html#method.to_int_unchecked)

Relevant tracking issues are here:

- https://internals.rust-lang.org/t/pre-rfc-add-explicitly-named-numeric-conversion-apis/11395
- https://github.com/rust-lang/rust/pull/66841
- https://github.com/rust-lang/rust/issues/67057
- https://github.com/rust-lang/rust/issues/67058